src/main/java/app/controller/soundEngine/CornerAimedTracing.java

```diff
@@ -5,41 +5,99 @@
+import java.util.HashMap;
+import java.util.LinkedHashSet;
 
 public class CornerAimedTracing implements SoundEngine{
-    public double compute(Map map, Agent agent){
-        ArrayList<SoundRay> soundRays = new ArrayList<>();
+    public HashMap<SoundSource, Sound> compute(Map map, Agent agent){
+        HashMap<SoundSource, Sound> soundTable = new HashMap<>();
+        int maxDiffraction = 1;
 
-        // Step 1: aim at all the soundsources
-        // right now this means ignore obstactles and return the loudest audible sound from agent
+        Vector agentPos = agent.getPosition();
 
-        double maxSoundLevel = 0.0;
+        ArrayList<Vector> points = new ArrayList<>();
+        points.add(agentPos);
 
-        for (SoundSource soundSource: map.getSoundSources()) {
-            double currentSoundLevel = soundSource.soundLevelFrom(agent.getPosition());
-            if(currentSoundLevel > maxSoundLevel) {
-                maxSoundLevel = currentSoundLevel;
+        ArrayList<Vector> nextToCheck = points;
+
+        for (int i = 0; i < maxDiffraction + 1; i++) {
+
+            ArrayList<Vector> corners = nextToCheck;
+            nextToCheck = new ArrayList<>();
+
+            for (Vector pos: corners) {
+                ArrayList<SoundSource> audible = audibleFrom(map, pos);
+
+                for (SoundSource heard: audible) {
+
+                    if(!soundTable.containsKey(heard)) {
+                        soundTable.put(heard, new Sound(heard, agentPos, pos, i));
+                        continue;
+                    }
+                    soundTable.get(heard).update(pos, i);
+                }
+
+                ArrayList<Vector> nextLayer = findReachableCorners(map, agentPos);
+
+                nextToCheck.addAll(nextLayer);
             }
         }
 
-        return maxSoundLevel;
+        return soundTable;
+    }
+
+    private ArrayList<SoundSource> audibleFrom(Map map, Vector pos){
+        ArrayList<SoundSource> audible = new ArrayList<>();
+
+        for (SoundSource soundSource: map.getSoundSources()) {
+            SoundRay rayToSource = new SoundRay(pos, soundSource.getPosition());
 
-        /*
-        for (SoundSource soundSource: map.getSoundSources()){
-            soundRays.add(new SoundRay(agent.getPosition(), soundSource.getPosition()));
+            if(!intersectsAnySoundFurniture(map.getSoundFurniture(), rayToSource)){
+                audible.add(soundSource);
+            }
         }
-         */
 
+        return audible;
+    }
+
+    private ArrayList<Vector> findReachableCorners(Map map, Vector pos){
+        // linked hashset, because order is important
+        LinkedHashSet<Vector> reachable = new LinkedHashSet<>();
 
-        // Step 2: for diffraction aim at all the corners
-        /*
+        // make sure that each corner can actually hear the source? so maybe store the loudest source in a var
         for (SoundFurniture soundFurniture: map.getSoundFurniture()) {
```

Comment:
Hey mate that's O(n^3) complexity, really expensive to be done every time step. Is there a possible refactor to lower the expense?

Reply:
To be fair, this complexity is high, but there is no other way to do this.  Just as with the ray-tracing checking for intersections (360 rays * n objects) is just one long for loop.

Reply:
It is n*m*l, but this just ends up being the total amount of corners. But I agree, this is not  although optimal yet since we can probably do a lot of pruning based on information we have about the corners. But this will have to do for now, and I will improve that in the next iteration.